### DISCUSSION TITLE
MPY: Can't mount to SD Card

### DISCUSSION THREAD

**Discussion by jkaminski9** [micropython/micropython#15820] [STM32 / Pyboard] [answered]
Good Afternoon,

I have a genuine PyBoard v1.1 running MP v1.22.2. 

I am unable to mount a Micro SD Card to this board. I receive the following error:

![image](https://github.com/user-attachments/assets/09983b0e-d37f-490c-9a00-989e72f2692d)

I am using a SanDisk Ultra 32GB MicroSD card. I have also tried with a 64GB MicroSD card and recieved the same issue. 

Been searching through the forums, and so far none of the solutions have been able to fix this problem. Any help would be appreciated. 

Thanks.

**Comment by peterhinch** [comment]
I have never seen such a message on a Pyboard. Have you checked that your SD cards have a normal filesystem when viewed from a PC?

**Reply by jkaminski9** [reply]
Both of the SD Cards have a normal file system. I've tried reformatting them as well, and ensured that they were FAT systems. 

One of the SD Cards had been previously used and the other was fresh out of the packaging.

**Reply by jkaminski9** [reply]
@peterhinch Here are a few more diagnostics. I have the default boot.py after factory resetting on the flash, and my main.py file is blank.

![image](https://github.com/user-attachments/assets/c22d31ce-e97a-464e-9c75-30ef5c7d6a0e)

**Reply by mendenm** [reply]
Don't you need to create the "/sd" directory before you mount on it?

**Reply by jkaminski9** [reply]
@mendenm Per the documentation I didn't believe I needed to. But I did try and got the same error, shown below:

![image](https://github.com/user-attachments/assets/102acfa6-62ba-426a-aa0f-85c29ed918e7)

**Reply by mendenm** [reply]
Wait, that looks as if your root filesystem is read-only.  You should have been able to create a new directory at "/".  
This may mean I do not understand the pyboard, since I have never used it.  I assume(d) it has some native flash that appears at boot, and you are mounting an auxiliary sd card.  Is the SD card the primary file system on this, not an add-on?

**Reply by mendenm** [reply]
have you tried 
```
os.mkdir("/flash/sd")
os.mount(sd,"/flash/sd")
```
and seeing if it works, with the sd card contents now at "/flash/sd".  I just noticed that the contents of the built-in flash is at "/flash", not at "/".  Maybe "/" isn't writeable on the pyboard filesystem.

**Reply by peterhinch** [reply]
I have been using Pyboards since the early days (ten years!). I have lost count of how many I have, with four in 24/7 use. Given a good card the on board SD interface just works. I've never needed to run diagnostics on them. If you power up a Pyboard with a card in place it should be mounted.

Note that (as far as I know) you can't fit a card with the power up.

I would install a fresh firmware build (just in case), and check power supply conditions (new cable, different USB power source). If it still fails, I would suspect a faulty Pyboard.

**Reply by jkaminski9** [reply]
@peterhinch Peter, thanks for looking into this for me. Appreciate your help.

I have 8 pyboards that I purchased from MicroPython's store, and all are exhibiting this issue. I'm going to assume that the issue lies with my SD cards. I will try ordering some new ones. Thank you again.

**Reply by jkaminski9** [reply]
@peterhinch I tried flashing a few different firmware versions, still the same error. I also loaded another brand new SD Card. I've even tried different USB cables / power sources. All give the same error. 

I looked at the board under a microscope and the connections to the SD Card adapter all seem fine. I really hope its not a faulty pyboard since I ordered 8 of them. But aside from this, they work as expected so I'm not sure they are faulty.

**Reply by robert-hh** [reply]
There have been discussions about SD cards not working. The old MicroPython forum at https://forum.micropython.org/ should have a few threads about it. What Size & Brand is the card you use. I have used cards with sizes between 8 and 64 GB all formatted with a single FAT32 partition. Some people posted having use 256GB cards.

**Reply by jkaminski9** [reply]
@robert-hh I am currently using a SanDisk 32GB, and a MicroHC 8GB. Both are setup as single partition FAT32. I have also tried a 64GB SanDisk with the same formatting. 

The thing that is odd to me is that I can seemingly talk to / read from the SDCard using pyb.SDCard().info() and pyb.SDCard.read(). I just cannot mount. 

I do not have any files loaded onto the SDCard, I just want to be able to see it when I run a listdir command.

Thank you.

**Reply by robert-hh** [reply]
> I just cannot mount.

With the STM32 port the SD card is already mounted during boot if present. So maybe you try to remount the SD card..

**Reply by jkaminski9** [reply]
@robert-hh See below for when I try to mount:

![image](https://github.com/user-attachments/assets/cde95890-637e-4046-b3de-985434c8f689)

Thank you for taking the time to help me.

**Reply by robert-hh** [reply]
So it looks as if the SD card is already mounted. What do you get with:

```
import os
print(os.getcwd())
```

I expect it to print /flash/sd, which is the mount point.

**Reply by jkaminski9** [reply, marked as answer]
@robert-hh I was able to get it to mount by executing the following:

```
MicroPython v1.22.2 on 2024-02-22; PYBv1.1 with STM32F405RG
Type "help()" for more information.
>>> import os
>>> sd = pyb.SDCard()
>>> os.VfsFat.mkfs(sd)
>>> os.mount(sd,'/sd')
>>> os.listdir('/')
['flash', 'sd']
```

Thank you to @robert-hh and @peterhinch for your patience and your help with this. I am just going to move this code into my boot.py file from now on.

**Reply by robert-hh** [reply]
> I am just going to move this code into my boot.py file from now on.

So it seems that formatting at the PC did not create a valid file system. `os.VfsFat.mkfs(sd)` (re-)formats the SD. So putting it into boot.py would erase the SD at every boot.

**Reply by jkaminski9** [reply]
@robert-hh Apologies, I misspoke. I don't have `os.VfsFat.mkfs(sd)` in my boot.py file.

I just have the code to mount it now. Which I only need if I have SKIPSD in my /flash directory, which I do for my application.

Sorry for the confusion.

**Comment by peterhinch** [comment]
I am still puzzled by this. On a Pyoard 1.1 with no `boot.py` on Flash or on the SD card, I get
```py
MicroPython v1.23.0 on 2024-06-02; PYBv1.1 with STM32F405RG
Type "help()" for more information.
>>> 
>>> import os
>>> os.listdir('/')
['flash', 'sd']
>>>  
```
I have never found it necessary to have explicit mounting code on a Pyboard.
